homeassistant/components/vesync/sensor.py

```diff
@@ -1,25 +1,125 @@
 """Support for power & energy sensors for VeSync outlets."""
+from collections.abc import Callable
+from dataclasses import dataclass
 import logging
 
 from homeassistant.components.sensor import (
     SensorDeviceClass,
     SensorEntity,
+    SensorEntityDescription,
     SensorStateClass,
 )
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import ENERGY_KILO_WATT_HOUR, POWER_WATT
+from homeassistant.const import (
+    CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+    ENERGY_KILO_WATT_HOUR,
+    PERCENTAGE,
+    POWER_WATT,
+)
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.dispatcher import async_dispatcher_connect
 from homeassistant.helpers.entity import EntityCategory
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
 
-from .common import VeSyncBaseEntity
+from .common import VeSyncBaseEntity, VeSyncDevice
 from .const import DOMAIN, VS_DISCOVERY, VS_SENSORS
+from .fan import SKU_TO_BASE_DEVICE
 from .switch import DEV_TYPE_TO_HA
 
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass
+class VeSyncSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[VeSyncDevice], StateType]
+
+
+@dataclass
+class VeSyncSensorEntityDescription(
+    SensorEntityDescription, VeSyncSensorEntityDescriptionMixin
+):
+    """Describe VeSync sensor entity."""
+
+    exists_fn: Callable[[VeSyncDevice], bool] = lambda _: True
+    update_fn: Callable[[VeSyncDevice], None] = lambda _: None
+
+
+def update_energy(device):
+    """Update outlet details and energy usage."""
+    device.update()
+    device.update_energy()
+
+
+def sku_from_device(device):
+    """Get the base device of which a device is an instance."""
+    return SKU_TO_BASE_DEVICE.get(device.device_type)
+
+
+def ha_dev_type(device):
+    """Get the homeassistant device_type for a given device."""
+    return DEV_TYPE_TO_HA.get(device.device_type)
+
+
+SENSORS: tuple[VeSyncSensorEntityDescription, ...] = (
+    VeSyncSensorEntityDescription(
+        key="filter-life",
+        name="Filter Life",
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda device: device.details["filter_life"],
+        exists_fn=lambda device: sku_from_device(device)
+        in [
```

Comment:
Please don't use lambdas that span more then one line. We could store the device names in a constant and pass them as a parameter to the exists function.